### DISCUSSION TITLE
how to update an attribute when the ORM inserts a row?

### DISCUSSION THREAD

**Discussion by gioxc88** [sqlalchemy/sqlalchemy#5932] [Usage Questions] [answered]
I have this very simple structure, 

`Employee` is the base class, `Engineer` is the subclass.
I want the bonus of the engineer to default to 50% of its salary.
But the salary is a column defined in the parent class `Employee`.

```

from sqlalchemy import Column, Integer, Float
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Employee(Base):
    __tablename__ = 'employee'
    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    type = Column(String(50))
    salary = Column(Float)


    __mapper_args__ = {
        'polymorphic_identity':'employee',
        'polymorphic_on':type
    }

def get_bonus(context):
    return context.get_current_parameters()['salary'] * 0.5


class Engineer(Employee):
    __tablename__ = 'engineer'
    id = Column(Integer, ForeignKey('employee.id'), primary_key=True)
    bonus = Column(Float default=get_bonus)

    __mapper_args__ = {
        'polymorphic_identity':'engineer',
    }

```

Now if I try to initialize `Engineer` like:

```
e = Engineer(enginee_name='Gio',  salary=1000)
```

I get a `sqlalchemy.exc.StatementError: (builtins.KeyError) 'salary'`
The problem is that the `context` doesn't have access to the parameters of the parent.
This is very weird. I expected that the `context` should have access to all the parameters.

**Comment by zzzeek** [comment, marked as answer]
hey there -

a column default is not the appropriate place to do this, as you noted, there's no "employee" data in there because that's a totally different table.   The column default handler is part of Core, not ORM, and Core knows nothing about table inheritance, that's an ORM feature.

the normal way to intercept objects before an insert proceeds is to use the [before_insert](https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=before_insert#sqlalchemy.orm.events.MapperEvents.before_insert) hook:

```python
@event.listens_for(Engineer, "before_insert")
def receive_before_insert(mapper, connection, target):
    if target.bonus is None and target.salary is not None:
        target.bonus = target.salary * 0.5

```

**Reply by gioxc88** [reply]
That was very helpful. The confusion arises because I am able to initialize the `Engineer` passing parameters of another table (`Employee`), so it only seems natural that the context has access to all the parameters passed to the class initialization.

That being said I was not aware of the event hooks. Thank you very much for pointing that out

**Reply by zzzeek** [reply]
SQLAlchemy has two distinct parts, the Core and the ORM.  The Core is very much a 1-1 mapping to SQL statements.    The ORM is an extra layer of abstraction on top that can do things such as have one Python class refer to more than one table.  But when ORM talks to Core, it has to emit two separate INSERT statements, one for each table.   SQL-level defaults are local to one statement at a time.
